revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,243 @@
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = &result.CertRevocationResult{
+		Result: result.ResultNonRevokable,
+		ServerResults: []*result.ServerResult{{
+			Result: result.ResultNonRevokable,
+			Error:  nil,
+		}},
+	}
+
+	wg.Wait()
+	return certResults, nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) *result.CertRevocationResult {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return &result.CertRevocationResult{
+			Result:        result.ResultNonRevokable,
+			ServerResults: []*result.ServerResult{toServerResult("", NoServerError{})},
+		}
+	}
+
+	serverResults := make([]*result.ServerResult, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		serverResult := checkStatusFromServer(cert, issuer, server, opts)
+		if serverResult.Result == result.ResultOK ||
+			serverResult.Result == result.ResultRevoked ||
+			(serverResult.Result == result.ResultUnknown && errors.Is(serverResult.Error, UnknownStatusError{})) {
+			// A valid response has been received from an OCSP server
+			// Result should be based on only this response, not any errors from
+			// other servers
+			return serverResultsToCertRevocationResult([]*result.ServerResult{serverResult})
+		}
+		serverResults[serverIndex] = serverResult
+	}
+	return serverResultsToCertRevocationResult(serverResults)
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) *result.ServerResult {
+	// Check valid server
+	if serverURL, err := url.Parse(server); err != nil || !strings.EqualFold(serverURL.Scheme, "http") {
+		// This function is only able to check servers that are accessible via HTTP
+		return toServerResult(server, GenericError{Err: fmt.Errorf("OCSPServer protocol %s is not supported", serverURL.Scheme)})
+	}
+
+	// Create OCSP Request
+	resp, err := executeOCSPCheck(cert, issuer, server, opts)
+	if err != nil {
+		// If there is a server error, attempt all servers before determining what to return
+		// to the user
+		return toServerResult(server, err)
+	}
+
+	// Validate OCSP response isn't expired
+	if time.Now().After(resp.NextUpdate) {
+		return toServerResult(server, GenericError{Err: errors.New("expired OCSP response")})
+	}
+
+	// Check for presence of pkix-ocsp-no-check and id-ce-invalidityDate
+	// extensions in response
+	foundNoCheck := false
+	foundInvalidityDate := false
+	var invalidityDate time.Time
+	for _, extension := range resp.Extensions {
+		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+			foundNoCheck = true
+		}
+		if !foundInvalidityDate && extension.Id.String() == invalidityDateOID {
+			foundInvalidityDate = true
+			rest, err := asn1.UnmarshalWithParams(extension.Value, &invalidityDate, "generalized")
+			if len(rest) > 0 || err != nil {
+				// invalidity date is invalid, treating as not present
+				foundInvalidityDate = false
+			}
+		}
+	}
```

Comment:
How about:
```suggestion
	for _, extension := range resp.Extensions {
		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
			foundNoCheck = true
		}
		if !foundInvalidityDate && extension.Id.String() == invalidityDateOID {
			rest, err := asn1.UnmarshalWithParams(extension.Value, &invalidityDate, "generalized")
			if len(rest) > 0 || err != nil {
				// invalidity date is invalid, treating as not present
				continue
			}
			foundInvalidityDate = true
		}
	}
```

Reply:
if SigningTime is zero then we dont need to perform computation of foundInvalidityDate and invalidityDate.

Reply:
Resolving because this logic is changed entirely in the new commit